### PR TITLE
Introduce delay before resizing tabs (fixes #465)

### LINKED ISSUE #465: Don't switch to large tabs (or small tabs) when my mouse is in the tabbar.
I'm fine with cheese being moved, I'm less fine with cheese being moved when I'm in the process of eating it 😄 

When a tab is closed, we should wait for the mouse to exit the tab area (if it is in it), and then perhaps wait for X more seconds of mouse-not-being-in-the-tab-area.

### PR DESCRIPTION
**Pull request by Manishearth**
Tabs will never be resized if the mouse is in the tab area.

When the mouse leaves the tab area, tabs will be resized after a short delay.

r? @bwinton

### CODE REVIEW

**Comment by ericawright** [comment]
inside of `resizetabs`, `mouseEntered` and `actuallyResizeTabs` functions this code is never used as the condition is never satisfied.
```
    if (this.resizeTimeout > 0) {
      this.window.clearTimeout(this.resizeTimeout);
      this.resizeTimeout = -1;
    }
```

File: verticaltabs.js [hunk 1]
```diff
@@ -554,7 +562,7 @@ VerticalTabs.prototype = {
     let tab_meta_image = document.getAnonymousElementByAttribute(aTab, 'anonid', 'tab-meta-image');
     tab_meta_image.style.backgroundImage = `url(moz-page-thumb://thumbnail/?url=${encodeURIComponent(this.getUri(aTab).spec)}), url(resource://tabcenter/skin/blank.png)`;
 
-    this.updateUriLabel(aTab);
+    //this.updateUriLabel(aTab);
```
Comment (ericawright) [inline verticaltabs.js]: can you remove this?

File: verticaltabs.js [hunk 2]
```diff
@@ -617,6 +636,40 @@ VerticalTabs.prototype = {
     }
   },
 
+  resizeTabs: function () {
```
Comment (ericawright) [inline verticaltabs.js]: `resizeTabs` and `actuallyResizeTabs` is confusing. Could we rename `resizeTabs` to something else?

**Comment by ericawright** [comment]
There's an inconsistency here with search, any ideas on this, or if it's acceptable @bwinton?
resize delay

File: verticaltabs.js [hunk 3]
```diff
@@ -617,6 +636,40 @@ VerticalTabs.prototype = {
...
+    if (this.resizeTimeout > 0) {
+      this.window.clearTimeout(this.resizeTimeout);
+      this.resizeTimeout = -1;
+    }
+    if (this.mouseInside) {
```
Comment (ericawright) [inline verticaltabs.js]: it seems that `this.needsResize` is not actually necessary and `this.mouseInside` does the job for it. I tried removing all references of `this.needsResize` without any problems.

File: verticaltabs.js [hunk 4]
```diff
@@ -55,6 +55,9 @@ Cu.import('resource://gre/modules/PageThumbs.jsm');
 const NS_XUL = 'http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul';
 const TAB_DROP_TYPE = 'application/x-moz-tabbrowser-tab';
 
+// Wait these many milliseconds before resizing tabs
+// after mousing out
+const WAIT_BEFORE_RESIZE = 2000;
```
Comment (ericawright) [inline verticaltabs.js]: 2 seconds seems like a bit long to me....

**Comment by Keith94** [comment]
It should also not resize while tab context menu is open, I think.

**Comment by bwinton** [comment]
I agree that typing in the search bar should resize the tabs (if necessary), no matter where the mouse is.  (But if you've searched for something are are closing tabs, they shouldn't resize until you move the mouse or type more.)

File: verticaltabs.js (hunk 2 above)
Comment (Popog) [inline verticaltabs.js]: It feels a new function isn't actually needed. This code would actually read cleaner if resize just used lambdas or bind to setup delay calls to itself.

Comment (Manishearth) [inline verticaltabs.js]: We need to call resizeTabs in many places though. I'd rather have that logic in one place.

**Comment by Manishearth** [comment]
Changed delay to 1s, made it run immediately on a filter. Filter now calls `actuallyResizeTabs()`.

**Comment by ericawright** [comment]
some of your changes haven't made it in (or at least the 1 s delay hasn't)

**Comment by Manishearth** [comment]
`git add` fail

**Comment by Manishearth** [comment]
Updated to fix merge conflicts

**Comment by bwinton** [comment]
Okay, I'm going to merge this, and we can deal with whatever else crops up in followup bugs.  Thanks Manish!

**Pull request merged**